File: c++/cote/GJM/원형_문자열.cpp
#include <iostream>
#include <string>
#include <random>
#define TEST_CASES 10
#define MAX_LENGTH 10
#define MIN_LENGTH 5

using namespace std;


string initialize(default_random_engine& gen){

	string s;
	uniform_int_distribution<int> dis('a', 'f');

	int n = gen() % (MAX_LENGTH - MIN_LENGTH +1) +MIN_LENGTH ;

	for(int i=0; i<n; i++)
		s += dis(gen);

	return s;
}

struct Comparator{

	vector<int>& group;
	int t;

	Comparator(vector<int>& group, int t) : group(group), t(t){}

	bool operator()(int a, int b){
		if(group[a] != group[b]) return group[a] < group[b];
		return group[a+t] < group[b+t];
	}

};


vector<int> getSuffixArray(const string& s){
	int n = s.size(), t=1;

	vector<int> group(n+1);
	for(int i=0; i<n; i++) group[i] = s[i];
	group[n] = -1;

	vector<int> perm(n,0);
	for(int i=0; i<n; i++) perm[i]= i;

	while(t < n){
		Comparator comp(group,t);
		sort(begin(perm), end(perm), comp);

		t *=2;
		if(t >= n) break;
		vector<int> newGroup(n+1);
		newGroup[n] = -1;
		newGroup[0] = 0;
		for(int i=1; i<n; i++){
			if(comp(perm[i-1], perm[i]))
				newGroup[perm[i]] = newGroup[perm[i-1]] +1;
			else 
				newGroup[perm[i]]= newGroup[perm[i-1]];
		}
		group = newGroup;
	}

	return perm;
}

string solve(const string& s){

	string s2 = s+s;
	vector<int> a = getSuffixArray(s2);

	for(int i=0; i<a.size(); i++)
		if(a[i] < s.size())
			return s2.substr(a[i], s.size());
	
	return "";
}

int main(void){

	default_random_engine gen(43);

	for(int it =0; it <TEST_CASES; it++){

		string target = initialize(gen);
		cout << target << " ";
		cout << solve(target) << endl;
	}

	return 0;
}

File: kotlin/src/main/kotlin/com/example/cote/GJM/원형_문자열.kt
package com.example.cote.GJM

import kotlin.random.*
import kotlin.Comparator

val MAX_LENGTH = 100
val MIN_LENGTH = 10

val TEST_CASE = 10

class Comp(
	val group: List<Int>,
	val t: Int
	) : Comparator<Int> {

	override fun compare(a: Int, b: Int): Int {
		if (group[a] != group[b]) {
			return group[a] - group[b]
		}
		return group[a + t] - group[b + t]
	} 
}


fun Random.nextChar(): Char = Char(nextInt('a'.code, 'c'.code))

fun initialize(random: Random): String {

	val n = random.nextInt(MAX_LENGTH - MIN_LENGTH) + MIN_LENGTH

	return buildString {
		repeat(n) {
			append(random.nextChar())
		}
	}
}

fun getSuffixArray(target: String): List<Int> {
	var n = target.length
	var t = 1
	var group = MutableList(n) {target[it].code} .apply {
		add(-1)
	}
	val perm = MutableList(n) {it}

	while (t < n) {
		val comp = Comp(group, t)
		perm.sortWith(comp)
		
		t *= 2
		val newgroup = MutableList(n) {0} .apply {
			repeat(t) {
				add(-1)
			}
		}

		for (i in 1..n-1) {
			if (comp.compare(i, i-1) < 0) {
				newgroup[perm[i]] = newgroup[perm[i - 1]] + 1
			} else {
				newgroup[perm[i]] = newgroup[perm[i - 1]]
			}
		}

		group = newgroup

	} 

	return perm
}

// 원형 문자열으로 만들 수 있는 문자열 중 사전순으로 가장 앞에 오는 문자열을 반환
fun solve(target: String): String {
	val newTarget = target + target
	val perm = getSuffixArray(newTarget)

	for (i in perm) {
		if (i < target.length) {
			return newTarget.substring(i, i + target.length)
		}
	}

	return ""
}

fun main() {
	val random = Random(43)

	repeat(TEST_CASE) {
		val target = initialize(random)
		println("${solve(target)}")
	}
}